samples/csharp/end-to-end-apps/MulticlassClassification-GitHubLabeler/GitHubLabeler/GitHubLabelerConsoleApp/Program.cs

```diff
@@ -0,0 +1,137 @@
+
+namespace GitHubLabeler
+{
+    internal static class Program
+    {
+        private static string AppPath => Path.GetDirectoryName(Environment.GetCommandLineArgs()[0]);
+
+        private static string BaseDatasetsLocation = @"../../../../Data";
+        private static string DataSetLocation = $"{BaseDatasetsLocation}/corefx-issues-train.tsv";
+        //private static string DataSetLocation => Path.Combine(AppPath, "Data", "corefx-issues-train.tsv");
+
+        private static string BaseModelsPath = @"../../../../MLModels";
+        private static string ModelFilePathName = $"{BaseModelsPath}/GitHubLabelerModel.zip";
+        //private static string ModelPath => Path.Combine(AppPath, "MLModels", "GitHubLabelerModel.zip");
+
+        public static IConfiguration Configuration { get; set; }
+        private static async Task Main(string[] args)
+        {
+            SetupAppConfiguration();
+
+            //1. ChainedBuilderExtensions and Train the model
+            BuildAndTrainModel(DataSetLocation, ModelFilePathName);
+
+            //2. Try/test to predict a label for a single hard-coded Issue
+            TestSingleLabelPrediction(ModelFilePathName);
+
+            //3. Predict Issue Labels and apply into a real GitHub repo
+            await PredictLabelsAndUpdateGitHub(ModelFilePathName);
+        }
+
+        public static void BuildAndTrainModel(string DataSetLocation, string ModelPath)
+        {
+            // Create MLContext to be shared across the model creation workflow objects 
+            // Set a random seed for repeatable/deterministic results across multiple trainings.
+            var mlContext = new MLContext(seed: 0);
+
+            // STEP 1: Common data loading configuration
+            DataLoader dataLoader = new DataLoader(mlContext);
+            var trainingDataView = dataLoader.GetDataView(DataSetLocation);
+
+            // STEP 2: Common data pre-process configuration with pipeline data transformations
+            var dataProcessor = new DataProcessor(mlContext);
+            var dataProcessPipeline = dataProcessor.DataProcessPipeline;
+
+            // (OPTIONAL) Peek data (such as 2 records) in training DataView after applying the ProcessPipeline's transformations into "Features" 
+            Common.ConsoleHelper.PeekDataViewInConsole<GitHubIssue>(mlContext, trainingDataView, dataProcessPipeline, 2);
+            //Common.ConsoleHelper.PeekVectorColumnDataInConsole(mlContext, "Features", trainingDataView, dataProcessPipeline, 2);
+
+            // STEP 3: Set the training algorithm, then create and config the modelBuilder                 
+            var trainer = mlContext.MulticlassClassification.Trainers.StochasticDualCoordinateAscent("Label", "Features");
+            var modelBuilder = new Common.ModelBuilder<GitHubIssue, GitHubIssuePrediction>(mlContext, dataProcessPipeline); //, trainer);
+            modelBuilder.AddTrainer(new SdcaMultiClassTrainer(mlContext, "Features", "Label"));
```

Comment:
Recommend `OVA-AveragedPerceptron{iter=10}` for a text problem. It's our general winner across many text datasets, and therefore is recommended in example code.

Reply:
Correct, it needs to be used in an OVA (one-vs-all) to upgrade from binary to multi-class. The OVA is the form it was benchmarked in.

Reply:
My point is that I don't see an "OVA-AveragedPerceptron" trainer in the `mlContext.MulticlassClassification.Trainers` catalog, here:

![image](https://user-images.githubusercontent.com/1712635/48017851-3a4b9f80-e0e4-11e8-8204-24baca2c2096.png)

Do you have any sample code for this approach using `OVA-AveragedPerceptron` for multi-class-classification?

Thanks! 👍

Reply:
Example of OVA-AveragedPerceptron: https://github.com/dotnet/machinelearning/blob/9d33efeed68e82ec67508117d239dfcdfa8121d9/test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamples.cs#L620-L624

If you use AP for text, be sure to set iter=10 -- default is 1 iteration, which I've been meaning to revisit as a topic. We discussed it internally a year ago and, if I recall, we decided to move AP to 10 iterations though haven't acted on it.